### PR TITLE
Save setting the same state in the Vertx local data duplicated context

### PR DESCRIPTION
**Pull request by franz1981**
This is part of a bigger problem shown by some benchmarks ie https://github.com/franz1981/quarkus-reactive-beer

Feel free to try it at home, it should be able to collect flamegraphs as well and *is not* I/O bound nor using a database (that means, how much is relevant what I've found, that's a fair and separate question)

Some evidence that is a worthy change:

image

In violet you can see the `VertxCurrentContext.set` calls, which set the same `io.quarkus.arc.impl.RequestContext$RequestContextState` over and over in the same context.
In my case, it happens a lot more often because it is filtering the elements (12) twice (I have observed ~50 identical `VertxCurrentContext.set`), while with a simpler one (which mimic what hibernate reactive would do while performing a query), just ~4. 
In both cases, avoiding performing `CHM::put` seems a great idea, given that would benefit card marking as mentioned already at https://github.com/quarkusio/quarkus/pull/30024

My concern now is that is now biased to make it pay more when such replacement happen instead, how much is it common? @mkouba ?

FYI the othe remaning low hanging fruit cost of context propagation is https://github.com/smallrye/smallrye-context-propagation/issues/424

**NOTE** 
this PR include some code refactoring in the caller class (performing activation) to remove from the methods the "slow paths" and makes smaller (in term of bytecode) each of the refactored methods: this would increase the chance to get inlined according to the OpenJDK inlining mechanism. It would be a great practice to always do like that, if won't harm the eyes of the readers (it shouldn't but is personal!).

### CODE REVIEW

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/RequestContext.java [hunk 1]
```diff
@@ -251,7 +264,7 @@ static class RequestContextState implements ContextState {
 
         RequestContextState(ConcurrentMap<Contextual<?>, ContextInstanceHandle<?>> value) {
             this.map = Objects.requireNonNull(value);
-            this.isValid = 1;
+            assert isValid == VALID;
```
Comment (mkouba) [inline independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/RequestContext.java:267]: Validation of the `assert` keyword is not always enabled, or?

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/RequestContext.java [hunk 2]
```diff
@@ -235,7 +247,8 @@ private ContextNotActiveException notActive() {
     }
 
     static class RequestContextState implements ContextState {
-
+        private static int VALID = 0;
```
Comment (mkouba) [inline independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/RequestContext.java:250]: Why not `final`?

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/RequestContext.java (hunk 1 above)
Comment (franz1981) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/RequestContext.java:267]: Nope, it should be enabled just on tests (with -ea), but I can removed it too; it's just to be sure that initial value won't change (like a self doc check)

Comment (Ladicek) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/RequestContext.java:267]: I think adding a comment would be enough, just to remind that the field is initialized to the default value of the type, which is `0`.

Comment (geoand) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/RequestContext.java:267]: Definitely +1 for a comment

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/RequestContext.java [hunk 3]
```diff
@@ -235,7 +247,8 @@ private ContextNotActiveException notActive() {
     }
 
     static class RequestContextState implements ContextState {
```
[the remark this answers is not in the document]
Comment (geoand) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/RequestContext.java:249]: I think something like this explanation needs to be in a comment somewhere otherwise we are bound to forget

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/RequestContext.java (hunk 1 above)
Comment (franz1981) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/RequestContext.java:267]: I'll go for it then, thanks!

File: extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java [hunk 4]
```diff
@@ -39,7 +39,13 @@ public void set(T state) {
             Context context = Vertx.currentContext();
             if (context != null && VertxContext.isDuplicatedContext(context)) {
                 VertxContextSafetyToggle.setContextSafe(context, true);
-                context.putLocal(LOCAL_KEY, state);
+                // this is racy but should be fine, because DC should not be shared
+                // and never remove the existing mapping
+                var oldState = context.getLocal(LOCAL_KEY);
+                if (oldState != state) {
+                    context.putLocal(LOCAL_KEY, state);
```
Comment (mkouba) [inline extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java:46]: Hm, I wonder if we shouldn't do this optimization on the `io.vertx.core.Context#putLocal()` level instead? I'm not against this modification but it looks like an implementation detail.

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/RequestContext.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (franz1981) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/RequestContext.java:249]: Adding a comment :+1: good point

File: independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/RequestContext.java (hunk 1 above)
Comment (franz1981) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/RequestContext.java:267]: Nope, not just for tests: I mean that assertions should be enabled there by default AFAIK, but is more a check for outself of the future in case we change the init semantic. But as said by others, comments can do a much better job then this, for humans

Comment (mkouba) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/RequestContext.java:267]: +1 for a comment ;-)

Comment (franz1981) [reply independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/RequestContext.java]: @mkouba how it looks like now? I've added a comment re the old JLS on java 8 as well, for completeness

**Review by mkouba — APPROVED** [review]

**Comment by franz1981** [comment]
Just noticed a thing

@mkouba 

```java

    @Override
    public void activate(ContextState initialState) {
        if (LOG.isTraceEnabled()) {
            traceActivate(initialState);
        }
        if (initialState == null) {
            currentContext.set(new RequestContextState(new ConcurrentHashMap<>()));
            // Fire an event with qualifier @Initialized(RequestScoped.class) if there are any observers for it
            fireIfNotEmpty(initializedNotifier);
        } else {
            if (initialState instanceof RequestContextState) {
                currentContext.set((RequestContextState) initialState);
            } else {
                throw new IllegalArgumentException("Invalid initial state: " + initialState.getClass().getName());
            }
        }
    }
    
```
on `RequestContext`, we end up on purpose to keep on adding it as far as I can read

**Comment by franz1981** [comment]
> In most applications the [quarkus.arc.context-propagation.enabled=false](https://quarkus.io/guides/cdi-reference#quarkus-arc_quarkus.arc.context-propagation.enabled) should be a safe performance boost (unless the SmallRye Context Propagation API is used explicitly).

That's awesome, do we have any docs on this for our users to explain it? Is it worthy?

**Comment by geoand** [comment]
Probably best to squash the commits

**Comment by franz1981** [comment]
Performance difference is pretty relevant I see; and I've removed smallrye decoration as well; getting another (small) performance boost.
I would still like to let common users (that won't disable that) to see a similar performance boost thanks to this PR and https://github.com/smallrye/smallrye-context-propagation/issues/424

File: extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java [hunk 5]
```diff
@@ -39,7 +39,13 @@ public void set(T state) {
             Context context = Vertx.currentContext();
             if (context != null && VertxContext.isDuplicatedContext(context)) {
                 VertxContextSafetyToggle.setContextSafe(context, true);
```
Comment (franz1981) [inline extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java:41]: @Sanne @mkouba I've already changed internally this via https://github.com/quarkusio/quarkus/pull/30024, but still...maybe it could be just avoided?

if `context.getLocal(LOCAL_KEY) == state` do we need to call `setContextSafe`? are they indipendent?

**Comment by geoand** [comment]
Where are we with this one? Is it ready to be merged? 

IIUC, we decided to not drop the ConcurrentHashMap because it could lead to subtle issues and we are not yet in a state where we can say that we explicitly don't support the cases that would break. Is that accurate?

**Comment by franz1981** [comment]
> Where are we with this one? Is it ready to be merged?

I need to rebase it and yep, we are ready to go!

>  Is that accurate?

Yep, we can talk about the future of such in a separate call, but in the current state avoiding wasteful CHM::put seems the easier thing to do ATM

**Comment by geoand** [comment]
👌

**Comment by geoand** [comment]
So unless there are any objections, let's merge this

File: extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java [hunk 6]
```diff
@@ -39,7 +39,13 @@ public void set(T state) {
...
-                context.putLocal(LOCAL_KEY, state);
+                // this is racy but should be fine, because DC should not be shared
+                // and never remove the existing mapping
```
Comment (Ladicek) [inline extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java:43]: I'm thinking about this and I was able to convince myself that it's true :-)

The only case when `oldState != state` could be when `oldState == null`. A single duplicated context can be easily shared between _threads_, but cannot be shared between _requests_, which is what matters here.

Comment (franz1981) [reply extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java:43]: I got some prior statements on the code (doesn't mean that is true eh): https://github.com/quarkusio/quarkus/blob/2e5c007ea2441ec9c4aaedc2e094f6a0dd4b966c/extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java#L52-L53

We should remove that one as well if we know that's not a valid statement.
In relation of the changes here (comment apart, that could be removed), if we have racy writes we're still at the mercy of casual ordering of acquisition of the the CHM segment "lock", which means...random.
In short, this PR shouldn't mess up with that, in theory.

Comment (Ladicek) [reply extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java:43]: No, that's true. Basically what I had to figure out for myself is what the comments mean when they say "should not be shared". It is entirely possible for the DCs to be shared among _threads_, but they should not be shared among _requests_. It probably took me longer than it should have, but it finally clicked :-)

Comment (geoand) [reply extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java:43]: > but they should not be shared among requests

Exactly true

Comment (franz1981) [reply extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java:43]: > It is entirely possible for the DCs to be shared among threads, but they should not be shared among requests.

But it still raise the question: 2 threads sharing such would end up competing to write different `state` instances for the same `LOCAL_KEY`? 
My answer is, probably not, if the first time is set/activated isn't racy, instead, while subsequent ones, yep.
The request ctx state than, should be the same on both threads who share it and it just means that my "racy check" would just save replacing the same value (which imply avoiding costy CHM::put over and over).

Comment (Ladicek) [reply extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java:43]: Right, that's also my conclusion. If 2 threads compete for writing to the same duplicated context, they should be both trying to write the same request context state.

Comment (franz1981) [reply extensions/vertx/runtime/src/main/java/io/quarkus/vertx/runtime/VertxCurrentContextFactory.java:43]: If that's true I would expect that is never racy at this point, because after the initial setting up, value will be stable (is not removed).
Now another question...

```
VertxContextSafetyToggle.setContextSafe(context, true);
```
This is still performing few checks if the context is duplicated bla bla...at this point I could shield calling it based on LOCAL_KEY's value being the same or not and save doing [these checks](https://github.com/quarkusio/quarkus/blob/53a64eaa2034375fcee588329b3450cbc41df5af/extensions/vertx/runtime/src/main/java/io/quarkus/vertx/core/runtime/context/VertxContextSafetyToggle.java#L119) over and over (although it is internally shielded, see https://github.com/quarkusio/quarkus/pull/30024)?

**Comment by franz1981** [comment]
@geoand @Ladicek how this look? Other things I could address on it?

**Comment by geoand** [comment]
+1 for merging

**Review by Ladicek — APPROVED** [review]
